area_detector_handlers/handlers.py

```diff
@@ -405,41 +406,64 @@ class IMMHandler(HandlerBase):
         number of frames to return as one datum
     """
     specs = {"IMM"} | HandlerBase.specs
+    return_type = {'delayed': True}
 
     def __init__(self, filename, frames_per_point):
-        self.file = open(filename, "rb")
-        self.frames_per_point = frames_per_point
-        header = read_imm_header(self.file)
-        self.rows, self.cols = header['rows'], header['cols']
-        self.is_compressed = bool(header['compression'] == 6)
-        self.file.seek(0)
-        self.toc = []  # (start byte, element count) pairs
-        while True:
-            try:
-                header = read_imm_header(self.file)
-                cur = self.file.tell()
-                payload_size = header['dlen'] * (6 if self.is_compressed else 2)
-                self.toc.append((cur, header['dlen']))
-                file_pos = payload_size + cur
-                self.file.seek(file_pos)
-                # Check for end of file.
-                if not self.file.read(4):
-                    break
-                self.file.seek(file_pos)
-            except Exception as err:
-                raise IOError("IMM file doesn't seems to be of right type") from err
+        self.filename = filename
+        with open(filename, "rb") as file:
+            self.frames_per_point = frames_per_point
+            header = read_imm_header(file)
+            self.rows, self.cols = header['rows'], header['cols']
+            self.is_compressed = bool(header['compression'] == 6)
+            file.seek(0)
+            self.toc = []  # (start byte, element count) pairs
+            while True:
+                try:
+                    header = read_imm_header(file)
+                    cur = file.tell()
+                    payload_size = header['dlen'] * (6 if self.is_compressed else 2)
+                    self.toc.append((cur, header['dlen']))
+                    file_pos = payload_size + cur
+                    file.seek(file_pos)
+                    # Check for end of file.
+                    if not file.read(4):
```

Comment:
Looks like this is validating that the file length is as reported by `payload_size` from the header. Not sure why this is necessary, but if the file length happened to be longer than stated wouldn't this be infinitely looping?

Reply:
Each trip through the loop reads one header and one payload. This isn’t validation; it’s looking ahead to see if there is another header’s worth of bytes to read. If so, we continue through the loop. Because each iteration advances through the file buffer, eventually we will reach the end and this condition will be met, exiting the loop. I don’t see a possibility of an infinite loop no matter what the file’s contents are because we will always eventually reach the end of the buffer if we don’t error out before then. Agree?

Reply:
Resolved in conversation on call.